The job **`x86_64-rust-for-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17078399193/job/48425778711) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48425778711) [(plain)](https://github.com/rust-lang/rust/commit/9fac587d06235b24a052d480e7a6cc77e902c8dc/checks/48425778711/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling libloading v0.8.8
error: expected `;`, found reserved identifier `$crate`
##[error]   --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libloading-0.8.8/src/os/unix/mod.rs:284:48
    |
270 | /         cfg_if::cfg_if! {
271 | |             // These targets are known to have MT-safe `dlerror`.
272 | |             if #[cfg(any(
273 | |                 target_os = "linux",
...   |
284 | |                 self.get_singlethreaded(symbol)
    | |                                                ^ help: add `;` here
...   |
288 | |         }
    | |         -
    | |         |
    | |_________unexpected token
    |           only `;` terminated statements or tail expressions are allowed after this attribute

error: expected `;`, found reserved identifier `$crate`
##[error]   --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libloading-0.8.8/src/os/unix/mod.rs:286:74
    |
270 | /         cfg_if::cfg_if! {
271 | |             // These targets are known to have MT-safe `dlerror`.
272 | |             if #[cfg(any(
273 | |                 target_os = "linux",
...   |
286 | |                 self.get_impl(symbol, || Err(crate::Error::DlSymUnknown))
    | |                                                                          ^ help: add `;` here
287 | |             }
288 | |         }
    | |         -
    | |         |
    | |_________unexpected token
    |           only `;` terminated statements or tail expressions are allowed after this attribute

error: could not compile `libloading` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to compile `bindgen-cli v0.65.1`, intermediate artifacts can be found at `/tmp/cargo-installdtrAWS`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
  local time: Tue Aug 19 19:24:14 UTC 2025
  network time: Tue, 19 Aug 2025 19:24:15 GMT
****************************************************************************
To find more information about this job, visit the following URL:
```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/rust-for-linux.html).

--- a/compiler/rustc_codegen_llvm/src/attributes.rs
+++ b/compiler/rustc_codegen_llvm/src/attributes.rs
@@ -436,6 +436,16 @@ pub(crate) fn llfn_attrs_from_instance<'ll, 'tcx>(
         || codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR_ZEROED)
     {
         to_add.push(create_alloc_family_attr(cx.llcx));
+        if let Some(zv) =
+            cx.tcx.get_attr(instance.def_id(), rustc_span::sym::rustc_allocator_zeroed_variant)
+            && let Some(name) = zv.value_str()
+        {
+            to_add.push(llvm::CreateAttrStringValue(
+                cx.llcx,
+                "alloc-variant-zeroed",
+                &mangle_internal_symbol(cx.tcx, name.as_str()),
+            ));
+        }
         // apply to argument place instead of function
         let alloc_align = AttributeKind::AllocAlign.create_attr(cx.llcx);
         attributes::apply_to_llfn(llfn, AttributePlace::Argument(1), &[alloc_align]);

--- a/compiler/rustc_feature/src/builtin_attrs.rs
+++ b/compiler/rustc_feature/src/builtin_attrs.rs
@@ -993,6 +993,10 @@ pub static BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
         rustc_allocator_zeroed, Normal, template!(Word), WarnFollowing,
         EncodeCrossCrate::No,
     ),
+    rustc_attr!(
+        rustc_allocator_zeroed_variant, Normal, template!(NameValueStr: "function"), ErrorPreceding,
+        EncodeCrossCrate::Yes,
+    ),
     gated!(
         default_lib_allocator, Normal, template!(Word), WarnFollowing,
         EncodeCrossCrate::No, allocator_internals, experimental!(default_lib_allocator),

--- a/compiler/rustc_span/src/symbol.rs
+++ b/compiler/rustc_span/src/symbol.rs
@@ -1816,6 +1816,7 @@ symbols! {
         rustc_align,
         rustc_allocator,
         rustc_allocator_zeroed,
+        rustc_allocator_zeroed_variant,
         rustc_allow_const_fn_unstable,
         rustc_allow_incoherent_impl,
         rustc_allowed_through_unstable_modules,

--- a/library/alloc/src/alloc.rs
+++ b/library/alloc/src/alloc.rs
@@ -17,6 +17,7 @@ unsafe extern "Rust" {
     #[rustc_allocator]
     #[rustc_nounwind]
     #[rustc_std_internal_symbol]
+    #[rustc_allocator_zeroed_variant = "__rust_alloc_zeroed"]
     fn __rust_alloc(size: usize, align: usize) -> *mut u8;
     #[rustc_deallocator]
     #[rustc_nounwind]

--- a/tests/codegen-llvm/vec-calloc.rs
+++ b/tests/codegen-llvm/vec-calloc.rs
@@ -1,4 +1,6 @@
+//@ revisions: normal llvm21
 //@ compile-flags: -Copt-level=3 -Z merge-functions=disabled
+//@ [llvm21] min-llvm-version: 21
 //@ only-x86_64
 
 #![crate_type = "lib"]
@@ -174,6 +176,24 @@ pub fn vec_option_i32(n: usize) -> Vec<Option<i32>> {
 
     // CHECK: ret void
     vec![None; n]
+}
+
+// LLVM21-LABEL: @vec_array
+#[cfg(llvm21)]
+#[no_mangle]
+pub fn vec_array(n: usize) -> Vec<[u32; 1_000_000]> {
+    // LLVM21-NOT: call {{.*}}alloc::vec::from_elem
+    // LLVM21-NOT: call {{.*}}reserve
+    // LLVM21-NOT: call {{.*}}__rust_alloc(
+
+    // LLVM21: call {{.*}}__rust_alloc_zeroed(
+
+    // LLVM21-NOT: call {{.*}}alloc::vec::from_elem
+    // LLVM21-NOT: call {{.*}}reserve
+    // LLVM21-NOT: call {{.*}}__rust_alloc(
+
+    // LLVM21: ret void
+    vec![[0; 1_000_000]; 3]
 }
 
 // Ensure that __rust_alloc_zeroed gets the right attributes for LLVM to optimize it away.

